distributed/protocol/serialize.py

```diff
@@ -0,0 +1,176 @@
+from __future__ import print_function, division, absolute_import
+
+from functools import partial
+
+from dask.base import normalize_token
+
+from . import pickle
+
+
+serializers = {}
+deserializers = {None: lambda header, frames: pickle.loads(b''.join(frames))}
+
+
+def register_serialization(cls, serialize, deserialize):
+    """ Register a new class for custom serialization
+
+    Parameters
+    ----------
+    cls: type
+    serialize: function
+    deserialize: function
+
+    Examples
+    --------
+    >>> class Human(object):
+    ...     def __init__(self, name):
+    ...         self.name = name
+
+    >>> def serialize(human):
+    ...     header = {}
+    ...     frames = [human.name.encode()]
+    ...     return header, frames
+
+    >>> def deserialize(header, frames):
+    ...     return Human(frames[0].decode())
+
+    >>> register_serialization(Human, serialize, deserialize)
+    >>> serialize(Human('Alice'))
+    ({}, [b'Alice'])
+
+    See Also
+    --------
+    serialize
+    deserialize
+    """
+    name = typename(cls)
+    serializers[name] = serialize
+    deserializers[name] = deserialize
+
+
+def typename(typ):
+    """ Return name of type
+
+    Examples
+    --------
+    >>> from distributed import Scheduler
+    >>> typename(Scheduler)
+    'distributed.scheduler.Scheduler'
+    """
+    return typ.__module__ + '.' + typ.__name__
+
+
+def serialize(x):
```

Comment:
I think I'm going to wait on this until it comes up in practice.  There are some tricky things to consider (how to avoid type checking all entries of long lists of text for example) and I'd like to ensure that what's in here handles itself well.